.github/workflows/scripts/inspect_remote_img.sh

```diff
@@ -1,24 +1,23 @@
 #!/bin/bash
 
 inspect_remote_img() {
-  if [[ $# -ne 2 ]]; then
-    echo 'inspect_remote_img $GH_TOKEN $IMAGE'
-    echo 'Example: inspect_remote_img XXXXXXXXXXXX ghcr.io/nvidia/t5x:latest'
+  if [[ $# -ne 4 ]]; then
+    echo 'inspect_remote_img $GH_TOKEN $IMAGE $OS $ARCH'
+    echo 'Example: inspect_remote_img XXXXXXXXXXXX ghcr.io/nvidia/upstream-t5x:latest linux amd64'
     echo 'Returns the metadata of a tagged remote image (no download)'
     return 1
   fi
   GH_TOKEN=$1
   IMAGE=$2
+  OS=$3
+  ARCH=$4
 
-  TAG=$(echo $IMAGE | rev | cut -d: -f1 | rev)
   IMAGE_REPO=$(echo $IMAGE | rev | cut -d: -f2- | rev)
 
   PACKAGE=$(echo $IMAGE_REPO | rev | cut -d/ -f1 | rev)
   ORG=$(echo $IMAGE_REPO | rev | cut -d/ -f2 | rev)
 
-  top_manifest_digest=$(curl -s -H "Authorization: Bearer $(echo $GH_TOKEN | base64)" "https://ghcr.io/v2/$ORG/$PACKAGE/manifests/$TAG" | jq -r .manifests[0].digest)
-  config_digest=$(curl -s -H "Authorization: Bearer $(echo $GH_TOKEN | base64)" "https://ghcr.io/v2/$ORG/$PACKAGE/manifests/$top_manifest_digest" | jq -r .config.digest)
-  img_metadata=$(curl -s -L -H "Authorization: Bearer $(echo $GH_TOKEN | base64)" "https://ghcr.io/v2/$ORG/$PACKAGE/blobs/$config_digest")
-  echo $img_metadata
+  config_digest=$(docker manifest inspect -v $IMAGE | jq -r ".[] | select(.Descriptor.platform == {\"architecture\": \"$ARCH\", \"os\": \"$OS\"}) | .OCIManifest.config.digest + .SchemaV2Manifest.config.digest")
+  curl -s -L -H "Authorization: Bearer $(echo $GH_TOKEN | base64)" "https://ghcr.io/v2/$ORG/$PACKAGE/blobs/$config_digest"
```

Comment:
Can we use the docker CLI instead of GitHub RESTful API for obtaining the digest here?

Reply:
This isn't obtaining the digest; it's getting the JSON configuration/metadata about the image
```
{
  "architecture": "amd64",
  "config": {
    "Env": [
      "PATH=/opt/amazon/efa/bin:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
      ...
    ],
    "Entrypoint": [
      "/opt/nvidia/nvidia_entrypoint.sh"
    ],
    "Labels": {
      "maintainer": "NVIDIA CORPORATION <cudatools@nvidia.com>",
      "org.opencontainers.image.created": "2023-10-16",
...
```
it wasn't clear how to get this metadata from the Docker CLI **without** downloading the image:
```
$ docker image inspect ghcr.io/nvidia/upstream-t5x:latest
[]
Error response from daemon: No such image: ghcr.io/nvidia/upstream-t5x:latest
$ docker pull ghcr.io/nvidia/upstream-t5x:latest
... slow fetch and extraction ...
$ docker image inspect ghcr.io/nvidia/upstream-t5x:latest
[
    {
        "Id": "sha256:ab3cda0001cfdbe163fe6587e7af726690100a25025591a9a42af50929e4584d",
        ...
        "Config": {
            ...
            "Env": [
                "PATH=/opt/amazon/efa/bin:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
...
```
the REST API is clearly one option that doesn't need new dependencies, and which is what we already did. Other tools like `skopeo` might also do the job, but we'd need to install that...

Reply:
The first snippet that you posted appears to be the output of `docker inspect`/`docker image inspect`.

[`docker manifest inspect`](https://docs.docker.com/engine/reference/commandline/manifest_inspect/) will not download the image and will instead query the manifest from the registry. The output will look like:
```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 529,
         "digest": "sha256:6f8686442577a7c61863e653f121aeddfc4e3624da1c308b761830e6dcbcc73f",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      ...
   ]
}
```

Reply:
Yes; that is the image digest:
```
$ docker manifest inspect ghcr.io/nvidia/upstream-t5x:latest | jq .manifests[0].digest
"sha256:4b63f505184003559014ad939912c8bb2cbc0629a3ebce8053ceeca76ec2c8e1"
$ docker manifest inspect -v ghcr.io/nvidia/upstream-t5x:latest | jq .[0].Descriptor.digest
"sha256:4b63f505184003559014ad939912c8bb2cbc0629a3ebce8053ceeca76ec2c8e1"
```
what the snippet is extracting is the config digest
```
$ docker manifest inspect -v ghcr.io/nvidia/upstream-t5x:latest | jq .[0].SchemaV2Manifest.config.digest
"sha256:ab3cda0001cfdbe163fe6587e7af726690100a25025591a9a42af50929e4584d"
```
and the `curl` command is downloading the config content using this digest. The point is to get information like
```
"org.opencontainers.image.created": "2023-10-16"
```
that is contained in the config content.

Reply:
Now I got it. I was initially confused because the script's name made me think it is a wrapper around `docker manifest inspect`. 🤣

Reply:
That said, `skopeo` is included by default in the `ubuntu-22.04` runner so why not we just use that?
I just used it in a recent PR: https://github.com/NVIDIA/JAX-Toolbox/blob/add-publish-single-arch-images/.github/workflows/_publish_container.yaml#L88

Reply:
RESTful API:
- pros: standardized (?), uses generic tooling
- cons: verbose

Skopeo:
- pros: simplifies many complicated docker operations
- cons: requires specialized software (available now in GH runners by default)

Reply:
What do you think? @olupton

Reply:
If we've already taken on the dependency on `skopeo` then that con doesn't really apply. So I guess we should use it here too. I'll take a look later.